protocol/x/vault/keeper/withdraw_test.go

```diff
@@ -0,0 +1,430 @@
+package keeper_test
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/cometbft/cometbft/types"
+	"github.com/dydxprotocol/v4-chain/protocol/dtypes"
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	testutil "github.com/dydxprotocol/v4-chain/protocol/testutil/util"
+	assettypes "github.com/dydxprotocol/v4-chain/protocol/x/assets/types"
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	pricestypes "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	satypes "github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+	vaulttypes "github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetVaultWithdrawalSlippagePpm(t *testing.T) {
+	testVaultId := constants.Vault_Clob1
+	testClobPair := constants.ClobPair_Eth
+	testPerpetual := constants.EthUsd_20PercentInitial_10PercentMaintenance
+	testMarketParam := constants.TestMarketParams[1]
+	testMarketPrice := constants.TestMarketPrices[1]
+
+	tests := map[string]struct {
+		/* --- Setup --- */
+		// skew.
+		skewFactorPpm uint32
+		// spread.
+		spreadMinPpm      uint32
+		spreadBufferPpm   uint32
+		minPriceChangePpm uint32
+		// leverage.
+		assetQuoteQuantums   *big.Int
+		positionBaseQuantums *big.Int
+		// function input.
+		vaultId              vaulttypes.VaultId
+		withdrawalPortionPpm *big.Int
+		/* --- Expectations --- */
+		expectedSlippagePpm *big.Int
+		expectedErr         string
+	}{
+		"Success: leverage 0, skew 2, spread 0.003, withdraw 10%": {
```

Comment:
nit: Add test where leverage ppm gets rounded down to 0?

Reply:
wouldn't that be testing leverage calculation instead of withdrawal?

Reply:
Yes, it would be. Is there a test for the leverage calculation where the fact that it's in ppm leads to the some small leverage being rounded down to 0?

Reply:
added a test case where `leverage * imf` gets below 6 decimal places and thus gets rounded up to `1 ppm`

Reply:
No test where leverage ppm is 0 despite position size being non-negative?

Reply:
changed leverage to rational number